### PR TITLE
variable-length-quantity: create test case generator

### PR DESCRIPTION
**Pull request by leenipper**
For #605

Generate two test case arrays, one for encode and
one for decode, from the canonical-data.json.

Update example solution to pass the tests.
The API changed to use slices in order to match
the canonical-data.json test cases.
Retain the old functions but rename to decodeInt and encodeInt
and utilize these as helpers to support the new slice based API.

Update test program to use generated test case arrays.
Put FAIL and PASS in test result output.

Increment testVersion and targetTestVersion in
example solution and test program.

### CODE REVIEW

File: exercises/variable-length-quantity/.meta/gen.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+type TestGroup struct {
+	Description string
+	Cases       []OneCase
+}
+
+type OneCase struct {
+	Description string
+	Property    string      // "encode" or "decode"
+	Input       []uint32    // supports both []byte and []uint32 in JSON.
+	Expected    interface{} // supports []byte, []uint32, or null in JSON.
```
Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: two thing stand out to me:

1. I was a bit surprised: I thought that since `nil` is assignable to `[]uint32` that `uint32` can be used here
2. I feel like the schema is not doing well for us. In the old way, where `encode` and `decode` are keys, we can have a different struct for each kind of case. But since now we can only tell by looking at `property`, we have to make our `OneCase` struct a union of all possible cases. Since you have written more generators than I have, I would ask you: How much does this bother you?

Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: > I thought that since nil is assignable to []uint32 that uint32 can be used here

Sorry I meant `[]uint32` of course.

Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: > How much does this bother you?

And I guess more importantly, does it bother enough that it's worth proposing a change?

Comment (leenipper) [inline exercises/variable-length-quantity/.meta/gen.go]: > []uint32 can be used here

Actually, I think you are right.  I jumped to the more general interface{}, likely because I had to previously for a different generator.

> In the old way, where encode and decode are keys, we can have a different struct for each kind of case.

This made it easier, I agree.

> does it bother enough that it's worth proposing a change?

Well, I've not thought much about what the change might be. But I do like simplicity, and the newer approach in the JSON is adding complexity to the generators. Perhaps there is a compromise.  Do you know the rational for the new way in the JSON ?

Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: > Do you know the rational for the new way in the JSON ?

I didn't know off the top of my head, so I went searching. The clearest explanation is probably: https://github.com/exercism/x-common/issues/336#issuecomment-279880708 : 

> In most of the test suites with more than one type of test, the test's type is encoded
in a property-key, in an object describing a test group.
> 
> There are two problem with that approach:
>
> * It mixes two different concepts regarding the tests: grouping and identification
> * It doesn't allow nesting of test groups. That would be nice to have, but is not really needed.
> * It doesn't allow grouping of test of different types, which would be really great.
> 
> Moving the **test type** near the **test data**, we solved all the above problem easily. It is theoretically sound and adds functionality.

I wonder if it would be too much to ask to have the keys again.

I'm wondering what other statically-typed language has generators. There's https://github.com/exercism/xscala/tree/master/testgen/src/main/scala , might need to take a look at what they're doing.

Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: and https://github.com/exercism/xocaml/tree/master/tools/test-generator

**Comment by leenipper** [comment]
I pushed an update which has a simpler generator.

File: exercises/variable-length-quantity/.meta/gen.go (hunk 1 above)
Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: I considered another alternative, which is... can we deal with this on the Go side without needing a change in x-common? I tried it in #677.

File: exercises/variable-length-quantity/example.go [hunk 2]
```diff
@@ -25,34 +29,76 @@ func EncodeVarint(x uint32) []byte {
...
 	var b byte
 	for n, b = range buf {
 		x = x << 7
-		x |= uint32(b) & 0x7F
+		x |= uint32(b) & 0x7f
 		if (b & 0x80) == 0 {
-			return x, n + 1
+			return x, n + 1, nil
+		} else if n == len(buf)-1 {
+			// error case
```
Comment (petertseng) [inline exercises/variable-length-quantity/example.go]: since this is the last iteration, this isn't necessary, is it?

Comment (petertseng) [inline exercises/variable-length-quantity/example.go]: yeah, still works if this is removed.

File: exercises/variable-length-quantity/.meta/gen.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
+// +build ignore
```
Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: we don't need to `build ignore` generators anymore, now that they're in the separate directory, right? They originally were to stop them from being caught when we `go test`, but now no longer necessary.

Doesn't need to be in this PR, maybe I'll make a separate one to remove them.

File: exercises/variable-length-quantity/example.go (hunk 2 above)
Comment (petertseng) [inline exercises/variable-length-quantity/example.go]: I just checked - the example solution still passes the tests even if these lines are removed, so I would suggest removing them.

Comment (leenipper) [inline exercises/variable-length-quantity/example.go]: Yes. Will remote those lines.

File: exercises/variable-length-quantity/.meta/gen.go (hunk 3 above)
Comment (leenipper) [inline exercises/variable-length-quantity/.meta/gen.go]: Ok, can be done with one sweeping commit on all the .meta/gen.go files easily later.

File: exercises/variable-length-quantity/.meta/gen.go (hunk 1 above)
Comment (leenipper) [inline exercises/variable-length-quantity/.meta/gen.go]: Yes, I've been looking at that one.

File: exercises/variable-length-quantity/.meta/gen.go (hunk 3 above)
Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: I'll make an issue so I don't forget (or if someone else wants to take it)

**Comment by leenipper** [comment]
Well, I did wonder about the 0-length array, whether it was clear enough.  Do you think it is worth adding an errorExpected bool for clarity in the test data ?  Will be fairly easy to do.

**Comment by petertseng** [comment]
I would add it - I'm looking at the example of https://github.com/exercism/xgo/blob/master/exercises/largest-series-product/cases_test.go where we could find out via `expected < 0` whether it's an error, but we still add an `ok` boolean anyway.

However, just to make sure I'm not falling to confirmation bias, I looked for a counterexample, and I found it in Forth: https://github.com/exercism/xgo/blob/master/exercises/forth/cases_test.go#L15 there is a comment there:

> // nil slice indicates error expected.

So it looks like there are two possible ways of making it clearer, so we have a choice.

**Comment by leenipper** [comment]
Ok. I chose to add the nil slice comment, and I updated test program to compare against nil as well.

**Comment by petertseng** [comment]
I just remembered something... forgot to check that `err` is an `error`? (#275)

Maybe we should only check it in the first few exercises that have `error` and not worry about it for the rest.

File: exercises/variable-length-quantity/.meta/gen.go (hunk 3 above)
Comment (petertseng) [inline exercises/variable-length-quantity/.meta/gen.go]: ok now that #682 is closed, gonna ask you to remove it from this one

**Comment by leenipper** [comment]
Accidental close earlier. Sorry. 
I updated the PR commit without the +build tag and with the var _ error check.
